### PR TITLE
Raise error when invalid attribute

### PR DESCRIPTION
**Pull request by GustavoCaso**
Solves #436 

Raises an error when invalid attribute name is defined by the SQL schema.

I did not know if this is the best place for raising this error. While I get some info I'm also looking for different places to raise such exception

### CODE REVIEW

**Comment by GustavoCaso** [comment]
This is the different configurations I used to reproduce the error

```ruby
config = ROM::Configuration.new(:sql, "postgresql://localhost/whitespace")


class Bugs < ROM::Relation[:sql]
  schema(infer: true)

  def index
    select { id.as(:uid) }
  end
end

class BugsRepository < ROM::Repository[:bugs]
end

config.register_relation Bugs

container = ROM.container(config)
connection = container.gateways[:default].connection
connection.execute "DELETE FROM bugs"
connection.execute "INSERT INTO bugs VALUES (1, 'WTF')"

repo = BugsRepository.new(container)

puts repo.bugs.to_a
```

```ruby
container = ROM.container(:sql, "postgresql://localhost/whitespace") do |config|
  config.relation(:bugs)
end

class BugsRepository < ROM::Repository[:bugs]
end

container.relations[:bugs].delete
container.relations[:bugs].insert("id" => 1, "white space" => "WTF")

BugsRepository.new(container).bugs.to_a
```

**Comment by solnic** [comment]
I believe this can be simply handled in `Schema#finalize_attributes`

**Comment by GustavoCaso** [comment]
You are right, after getting my hand in the finalize process and got to the same conclusion.

I will update the code.

Thanks for the hint.

**Comment by GustavoCaso** [comment]
@solnic after checking the code and making the test pass. I'm thinking maybe we could move the check to `Inferrer#call` https://github.com/rom-rb/rom/blob/master/core/lib/rom/schema/inferrer.rb#L48 so we remove the `if` from the code

**Comment by GustavoCaso** [comment]
@solnic do you like this approach better?

**Comment by solnic** [comment]
Thanks! I'm gonna merge it in but then I'll follow up in master with some minor tweaks.

**Comment by GustavoCaso** [comment]
@solnic the work that is missing is related to users using `alias`?

**Comment by GustavoCaso** [comment]
@solnic should we continue working on this one? If you point me to the work that is missing I would love to finish with it.

**Comment by solnic** [comment]
@GustavoCaso 👋 OK let's try this one more time 😉 to fix this, all we need to do is to raise the error [right here](https://github.com/rom-rb/rom/blob/master/core/lib/rom/struct_compiler.rb#L73-L74) when `name` includes a white-space.

File: core/lib/rom/schema.rb [hunk 1]
```diff
@@ -62,6 +62,8 @@ class Schema
 
     DEFAULT_INFERRER = Inferrer.new(enabled: false).freeze
 
+    VALID_IVAR_REGEX = /^[a-z_][a-zA-Z_0-9]*$/
```
Comment (gmkohler) [inline core/lib/rom/schema.rb]: Why must the instance variable have a lowercase letter first?  This seems to work fine:

```rb
2.6.3 :001 > Object.new.instance_variable_set :@A, :foo
 => :foo
2.6.3 :002 > "A".match( /^[a-z_][a-zA-Z_0-9]*$/)
 => nil
```

Would it make sense to catch the `NameError` raised by Ruby instead?
```rb
2.6.3 :005 > Object.new.instance_variable_set :"@A bad choice", :foo
Traceback (most recent call last):
        6: from /Users/gregorykohler/.rvm/rubies/ruby-2.6.3/bin/irb:23:in `<main>'
        5: from /Users/gregorykohler/.rvm/rubies/ruby-2.6.3/bin/irb:23:in `load'
        4: from /Users/gregorykohler/.rvm/rubies/ruby-2.6.3/lib/ruby/gems/2.6.0/gems/irb-1.0.0/exe/irb:11:in `<top (required)>'
        3: from (irb):5
        2: from (irb):5:in `rescue in irb_binding'
        1: from (irb):5:in `instance_variable_set'
NameError (`@A bad choice' is not allowed as an instance variable name)
```

(I haven't found where in the C code the validity is defined)

**Comment by solnic** [comment]
this needs to be reworked, so I'm closing this PR

**Pull request closed without merging**